File: secure_api_client.js
#!/usr/bin/env node

/**
 * Secure ERPNext API Client - Node.js/Axios Version
 * Demonstrates best practices for API authentication and security
 */

const axios = require('axios');
const fs = require('fs').promises;
const readline = require('readline');
const crypto = require('crypto');
const { URL } = require('url');

class ERPNextSecureClient {
    constructor(baseUrl = 'http://localhost:8080') {
        this.baseUrl = baseUrl.replace(/\/$/, ''); // Remove trailing slash
        this.authMethod = null;
        this.currentUser = 'unknown';
        
        // Create axios instance with security defaults
        this.client = axios.create({
            baseURL: this.baseUrl,
            timeout: 30000, // 30 second timeout
            headers: {
                'User-Agent': 'ERPNext-Secure-Client-JS/1.0',
                'Accept': 'application/json',
                'Content-Type': 'application/json'
            }
        });

        // Security check
        const url = new URL(this.baseUrl);
        if (url.protocol === 'http:' && !url.hostname.includes('localhost')) {
            console.warn('⚠️  WARNING: Using HTTP with non-localhost. Use HTTPS in production!');
        }

        // Add request/response interceptors for security
        this.setupInterceptors();
    }

    setupInterceptors() {
        // Request interceptor - add security headers and logging
        this.client.interceptors.request.use(
            (config) => {
                // Add timestamp for audit
                config.headers['X-Request-Time'] = new Date().toISOString();
                
                // Add request ID for tracing
                config.headers['X-Request-ID'] = this.generateRequestId();
                
                return config;
            },
            (error) => {
                this.logRequest('REQUEST_ERROR', '', 0, error.message);
                return Promise.reject(error);
            }
        );

        // Response interceptor - handle auth errors and logging
        this.client.interceptors.response.use(
            (response) => {
                // Log successful requests
                this.logRequest(
                    response.config.method.toUpperCase(),
                    response.config.url,
                    response.status
                );
                return response;
            },
            (error) => {
                const status = error.response?.status || 0;
                const method = error.config?.method?.toUpperCase() || 'UNKNOWN';
                const url = error.config?.url || '';

                // Handle specific error cases
                if (status === 401) {
                    console.error('❌ Authentication failed. Token may be expired.');
                } else if (status === 403) {
                    console.error('❌ Access forbidden. Check permissions.');
                } else if (status === 429) {
                    console.error('❌ Rate limit exceeded. Please wait.');
                }

                this.logRequest(method, url, status, error.message);
                return Promise.reject(error);
            }
        );
    }

    generateRequestId() {
        return crypto.randomBytes(8).toString('hex');
    }

    /**
     * Login using username/password (creates session cookie)
     * SECURITY: Use only for web applications, not for API clients
     */
    async loginWithCredentials(username, password) {
        if (!username) {
            username = await this.promptInput('Username: ');
        }
        if (!password) {
            password = await this.promptPassword('Password: ');
        }

        const loginData = { usr: username, pwd: password };

        try {
            const response = await this.client.post('/api/method/login', loginData);
            
            if (response.data.message && response.data.message.includes('Logged In')) {
                this.authMethod = 'session';
                this.currentUser = username;
                console.log('✅ Logged in successfully (session-based)');
                this.logAuthEvent('LOGIN_SUCCESS', username);
                return true;
            } else {
                console.log('❌ Login failed');
                this.logAuthEvent('LOGIN_FAILED', username);
                return false;
            }
        } catch (error) {
            console.error(`❌ Login error: ${error.message}`);
            this.logAuthEvent('LOGIN_ERROR', username, error.message);
            return false;
        }
    }

    /**
     * Setup token-based authentication
     * SECURITY: Recommended for API clients and server-to-server communication
     */
    async authenticateWithToken(apiKey, apiSecret) {
        if (!apiKey) {
            apiKey = process.env.ERPNEXT_API_KEY;
            if (!apiKey) {
                apiKey = await this.promptInput('API Key: ');
            }
        }

        if (!apiSecret) {
            apiSecret = process.env.ERPNEXT_API_SECRET;
            if (!apiSecret) {
                apiSecret = await this.promptPassword('API Secret: ');
            }
        }

        this.apiKey = apiKey;
        this.apiSecret = apiSecret;
        this.authMethod = 'token';

        // Set authorization header
        this.client.defaults.headers.common['Authorization'] = `token ${apiKey}:${apiSecret}`;

        // Test the token
        try {
            await this.get('/api/resource/User', { limit_page_length: 1 });
            console.log('✅ Token authentication successful');
            this.logAuthEvent('TOKEN_AUTH_SUCCESS', `${apiKey.substring(0, 8)}...`);
            return true;
        } catch (error) {
            console.error(`❌ Token authentication failed: ${error.message}`);
            this.logAuthEvent('TOKEN_AUTH_FAILED', `${apiKey.substring(0, 8)}...`, error.message);
            return false;
        }
    }

    generateApiKeyInstructions() {
        console.log('\n' + '='.repeat(60));
        console.log('HOW TO GENERATE API KEYS:');
        console.log('='.repeat(60));
        console.log('1. Login to ERPNext web interface');
        console.log('2. Go to Settings → My Settings');
        console.log('3. Scroll to \'API Access\' section');
        console.log('4. Click \'Generate Keys\'');
        console.log('5. Copy the API Key and API Secret');
        console.log('6. Store them securely (environment variables recommended)');
        console.log('\nEnvironment Variables:');
        console.log('export ERPNEXT_API_KEY=\'your_api_key_here\'');
        console.log('export ERPNEXT_API_SECRET=\'your_api_secret_here\'');
        console.log('\nNode.js (.env file):');
        console.log('ERPNEXT_API_KEY=your_api_key_here');
        console.log('ERPNEXT_API_SECRET=your_api_secret_here');
        console.log('='.repeat(60));
    }

    /**
     * Secure API Methods
     */
    async get(endpoint, params = {}) {
        this.checkAuth();
        const response = await this.client.get(endpoint, { params });
        return response.data;
    }

    async post(endpoint, data = {}) {
        this.checkAuth();
        const response = await this.client.post(endpoint, data);
        return response.data;
    }

    async put(endpoint, data = {}) {
        this.checkAuth();
        const response = await this.client.put(endpoint, data);
        return response.data;
    }

    async delete(endpoint) {
        this.checkAuth();
        const response = await this.client.delete(endpoint);
        return response.data;
    }

    checkAuth() {
        if (!this.authMethod) {
            throw new Error('Not authenticated. Use loginWithCredentials() or authenticateWithToken()');
        }
    }

    async logout() {
        if (this.authMethod === 'session') {
            try {
                await this.client.post('/api/method/logout');
                console.log('✅ Logged out successfully');
            } catch (error) {
                // Ignore logout errors
            }
        }

        // Clear auth headers and reset state
        delete this.client.defaults.headers.common['Authorization'];
        this.authMethod = null;
        this.currentUser = 'unknown';
        console.log('🔒 Session cleared');
    }

    /**
     * Utility Methods
     */
    async promptInput(question) {
        const rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });

        return new Promise((resolve) => {
            rl.question(question, (answer) => {
                rl.close();
                resolve(answer);
            });
        });
    }

    async promptPassword(question) {
        const rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });

        return new Promise((resolve) => {
            rl.question(question, (answer) => {
                rl.close();
                resolve(answer);
            });
            rl.stdoutMuted = true;
            rl._writeToOutput = function _writeToOutput(stringToWrite) {
                if (rl.stdoutMuted) {
                    rl.output.write('*');
                } else {
                    rl.output.write(stringToWrite);
                }
            };
        });
    }

    async logAuthEvent(event, user, details = '') {
        const timestamp = new Date().toISOString();
        const logEntry = `${timestamp} - ${event} - User: ${user} - ${details}\n`;
        
        try {
            await fs.appendFile('api_security.log', logEntry);
        } catch (error) {
            // Don't fail if logging fails
        }
    }

    async logRequest(method, endpoint, statusCode, error = '') {
        const timestamp = new Date().toISOString();
        const logEntry = `${timestamp} - ${method} ${endpoint} - ${statusCode} - User: ${this.currentUser} - ${error}\n`;
        
        try {
            await fs.appendFile('api_requests.log', logEntry);
        } catch (error) {
            // Don't fail if logging fails
        }
    }
}

/**
 * Advanced Secure Client with additional features
 */
class ERPNextAdvancedSecureClient extends ERPNextSecureClient {
    constructor(baseUrl, options = {}) {
        super(baseUrl);
        
        this.options = {
            retryAttempts: options.retryAttempts || 3,
            retryDelay: options.retryDelay || 1000,
            enableCache: options.enableCache || false,
            cacheTimeout: options.cacheTimeout || 300000, // 5 minutes
            ...options
        };

        this.cache = new Map();
        this.setupRetryLogic();
        
        if (options.rateLimitPerMinute) {
            this.setupRateLimit(options.rateLimitPerMinute);
        }
    }

    setupRetryLogic() {
        // Add retry interceptor
        this.client.interceptors.response.use(
            response => response,
            async (error) => {
                const config = error.config;

                // Don't retry if we've exceeded max attempts or it's not a retriable error
                if (!config || config.retry >= this.options.retryAttempts) {
                    return Promise.reject(error);
                }

                // Only retry on network errors or 5xx status codes
                const shouldRetry = !error.response || 
                    (error.response.status >= 500 && error.response.status < 600);

                if (!shouldRetry) {
                    return Promise.reject(error);
                }

                config.retry = (config.retry || 0) + 1;
                
                console.log(`🔄 Retrying request (${config.retry}/${this.options.retryAttempts}): ${config.url}`);
                
                // Wait before retrying
                await new Promise(resolve => setTimeout(resolve, this.options.retryDelay * config.retry));
                
                return this.client(config);
            }
        );
    }

    setupRateLimit(requestsPerMinute) {
        this.rateLimitQueue = [];
        this.rateLimitWindow = 60000; // 1 minute
        this.maxRequests = requestsPerMinute;

        this.client.interceptors.request.use(async (config) => {
            await this.checkRateLimit();
            return config;
        });
    }

    async checkRateLimit() {
        const now = Date.now();
        
        // Remove old requests outside the window
        this.rateLimitQueue = this.rateLimitQueue.filter(
            timestamp => now - timestamp < this.rateLimitWindow
        );

        // Check if we're at the limit
        if (this.rateLimitQueue.length >= this.maxRequests) {
            const oldestRequest = Math.min(...this.rateLimitQueue);
            const waitTime = this.rateLimitWindow - (now - oldestRequest);
            
            console.log(`⏳ Rate limit reached. Waiting ${Math.ceil(waitTime / 1000)}s...`);
            await new Promise(resolve => setTimeout(resolve, waitTime));
        }

        this.rateLimitQueue.push(now);
    }

    async get(endpoint, params = {}) {
        // Check cache first
        if (this.options.enableCache) {
            const cacheKey = `GET:${endpoint}:${JSON.stringify(params)}`;
            const cached = this.cache.get(cacheKey);
            
            if (cached && Date.now() - cached.timestamp < this.options.cacheTimeout) {
                console.log(`💾 Cache hit: ${endpoint}`);
                return cached.data;
            }
        }

        const result = await super.get(endpoint, params);

        // Cache the result
        if (this.options.enableCache) {
            const cacheKey = `GET:${endpoint}:${JSON.stringify(params)}`;
            this.cache.set(cacheKey, {
                data: result,
                timestamp: Date.now()
            });
        }

        return result;
    }

    clearCache() {
        this.cache.clear();
        console.log('🗑️  Cache cleared');
    }
}

/**
 * Demo Functions
 */
async function demoSecureUsage() {
    console.log('ERPNext Secure API Client Demo (Node.js/Axios)');
    console.log('='.repeat(50));

    const client = new ERPNextSecureClient();

    // Method 1: API Token (Recommended for APIs)
    console.log('\n🔐 Method 1: API Token Authentication (Recommended)');
    console.log('-'.repeat(50));

    const success = await client.authenticateWithToken();
    
    if (success) {
        try {
            console.log('\n📊 Fetching system info...');
            const systemSettings = await client.get('/api/resource/System%20Settings/System%20Settings');
            if (systemSettings?.data) {
                console.log(`   Country: ${systemSettings.data.country || 'Not set'}`);
                console.log(`   Time Zone: ${systemSettings.data.time_zone || 'Not set'}`);
            }

            console.log('\n👥 Fetching users (limited)...');
            const users = await client.get('/api/resource/User', { limit_page_length: 3 });
            if (users?.data) {
                users.data.forEach(user => {
                    console.log(`   - ${user.full_name || 'Unknown'} (${user.name || 'unknown'})`);
                });
            }

            console.log('\n🏢 Checking companies...');
            const companies = await client.get('/api/resource/Company');
            if (companies?.data) {
                companies.data.forEach(company => {
                    console.log(`   - ${company.name || 'Unknown Company'}`);
                });
            }

            // Demo creating a customer (if you have permissions)
            console.log('\n👤 Demo: Creating a test customer...');
            try {
                const newCustomer = await client.post('/api/resource/Customer', {
                    customer_name: 'Test Customer JS',
                    customer_type: 'Individual',
                    customer_group: 'All Customer Groups',
                    territory: 'All Territories'
                });
                console.log(`   ✅ Created customer: ${newCustomer.data.name}`);
            } catch (error) {
                console.log(`   ℹ️  Skipping customer creation (permission/validation error)`);
            }

        } catch (error) {
            console.error(`❌ Error during API calls: ${error.message}`);
        }

        await client.logout();
    } else {
        client.generateApiKeyInstructions();
    }
}

async function demoAdvancedFeatures() {
    console.log('\n🚀 Advanced Features Demo');
    console.log('-'.repeat(50));

    const advancedClient = new ERPNextAdvancedSecureClient('http://localhost:8080', {
        retryAttempts: 3,
        retryDelay: 1000,
        enableCache: true,
        cacheTimeout: 60000, // 1 minute cache
        rateLimitPerMinute: 30
    });

    if (await advancedClient.authenticateWithToken()) {
        console.log('\n💾 Testing caching...');
        
        // First request (will be cached)
        console.time('First request');
        await advancedClient.get('/api/resource/User', { limit_page_length: 1 });
        console.timeEnd('First request');
        
        // Second request (from cache)
        console.time('Cached request');
        await advancedClient.get('/api/resource/User', { limit_page_length: 1 });
        console.timeEnd('Cached request');
        
        advancedClient.clearCache();
        await advancedClient.logout();
    }
}

function printSecurityRecommendations() {
    console.log('\n' + '='.repeat(60));
    console.log('🔒 SECURITY RECOMMENDATIONS');
    console.log('='.repeat(60));
    console.log('1. ✅ USE API TOKENS for server-to-server communication');
    console.log('2. ✅ USE HTTPS in production (never HTTP)');
    console.log('3. ✅ STORE credentials in environment variables (.env)');
    console.log('4. ✅ IMPLEMENT rate limiting and retry logic');
    console.log('5. ✅ LOG all API access for audit trails');
    console.log('6. ✅ VALIDATE all inputs and handle errors gracefully');
    console.log('7. ✅ USE request timeouts and proper error handling');
    console.log('8. ✅ IMPLEMENT caching for frequently accessed data');
    console.log('9. ✅ MONITOR API usage and performance metrics');
    console.log('10. ✅ ROTATE API keys regularly (every 90 days)');
    console.log('\n❌ AVOID:');
    console.log('- Never commit API keys to version control');
    console.log('- Never use HTTP in production');
    console.log('- Never ignore SSL certificate errors');
    console.log('- Never expose API keys in client-side code');
    console.log('- Never log sensitive data');
    console.log('='.repeat(60));

    console.log('\n📦 Required Dependencies:');
    console.log('npm install axios dotenv');
    console.log('\n📄 Example .env file:');
    console.log('ERPNEXT_API_KEY=your_api_key_here');
    console.log('ERPNEXT_API_SECRET=your_api_secret_here');
    console.log('ERPNEXT_URL=https://your-domain.com');
}

// Main execution
async function main() {
    try {
        // Load environment variables if available
        try {
            require('dotenv').config();
        } catch (error) {
            console.log('💡 Tip: Install dotenv for .env file support: npm install dotenv');
        }

        await demoSecureUsage();
        await demoAdvancedFeatures();
        printSecurityRecommendations();
    } catch (error) {
        if (error.code === 'ECONNREFUSED') {
            console.error('\n❌ Connection refused. Make sure ERPNext is running on http://localhost:8080');
        } else {
            console.error(`\n❌ Unexpected error: ${error.message}`);
        }
    }
}

// Export classes for use as modules
module.exports = {
    ERPNextSecureClient,
    ERPNextAdvancedSecureClient
};

// Run demo if called directly
if (require.main === module) {
    main().catch(error => {
        console.error('Fatal error:', error.message);
        process.exit(1);
    });
}

File: secure_api_client.py
#!/usr/bin/env python3
"""
Secure ERPNext API Client
Demonstrates best practices for API authentication and security
"""

import os
import requests
import json
import getpass
from datetime import datetime
import hashlib
import hmac
from urllib.parse import urlparse

class ERPNextSecureClient:
    """
    Secure ERPNext API Client with multiple authentication methods
    """
    
    def __init__(self, base_url="http://localhost:8080"):
        self.base_url = base_url.rstrip('/')
        self.session = requests.Session()
        self.auth_method = None
        
        # Security headers
        self.session.headers.update({
            'User-Agent': 'ERPNext-Secure-Client/1.0',
            'Accept': 'application/json',
            'Content-Type': 'application/json'
        })
        
        # Verify SSL in production
        if urlparse(base_url).scheme == 'https':
            self.session.verify = True
        else:
            print("⚠️  WARNING: Using HTTP. Use HTTPS in production!")
    
    def login_with_credentials(self, username=None, password=None):
        """
        Login using username/password (creates session cookie)
        SECURITY: Use only for web applications, not for API clients
        """
        if not username:
            username = input("Username: ")
        if not password:
            password = getpass.getpass("Password: ")
        
        login_data = {"usr": username, "pwd": password}
        
        try:
            response = self.session.post(
                f"{self.base_url}/api/method/login",
                json=login_data
            )
            response.raise_for_status()
            
            result = response.json()
            if "message" in result and "Logged In" in result["message"]:
                self.auth_method = "session"
                print("✅ Logged in successfully (session-based)")
                self._log_auth_event("LOGIN_SUCCESS", username)
                return True
            else:
                print("❌ Login failed")
                self._log_auth_event("LOGIN_FAILED", username)
                return False
                
        except requests.exceptions.RequestException as e:
            print(f"❌ Login error: {e}")
            self._log_auth_event("LOGIN_ERROR", username, str(e))
            return False
    
    def authenticate_with_token(self, api_key=None, api_secret=None):
        """
        Setup token-based authentication
        SECURITY: Recommended for API clients and server-to-server communication
        """
        if not api_key:
            api_key = os.environ.get('ERPNEXT_API_KEY')
            if not api_key:
                api_key = input("API Key: ")
        
        if not api_secret:
            api_secret = os.environ.get('ERPNEXT_API_SECRET')
            if not api_secret:
                api_secret = getpass.getpass("API Secret: ")
        
        self.api_key = api_key
        self.api_secret = api_secret
        self.auth_method = "token"
        
        # Update session headers for token auth
        self.session.headers.update({
            'Authorization': f'token {api_key}:{api_secret}'
        })
        
        # Test the token
        try:
            response = self.get('/api/resource/User', params={'limit_page_length': 1})
            print("✅ Token authentication successful")
            self._log_auth_event("TOKEN_AUTH_SUCCESS", api_key[:8] + "...")
            return True
        except Exception as e:
            print(f"❌ Token authentication failed: {e}")
            self._log_auth_event("TOKEN_AUTH_FAILED", api_key[:8] + "...", str(e))
            return False
    
    def generate_api_key_instructions(self):
        """
        Print instructions for generating API keys
        """
        print("\n" + "="*60)
        print("HOW TO GENERATE API KEYS:")
        print("="*60)
        print("1. Login to ERPNext web interface")
        print("2. Go to Settings → My Settings")
        print("3. Scroll to 'API Access' section")
        print("4. Click 'Generate Keys'")
        print("5. Copy the API Key and API Secret")
        print("6. Store them securely (environment variables recommended)")
        print("\nEnvironment Variables:")
        print("export ERPNEXT_API_KEY='your_api_key_here'")
        print("export ERPNEXT_API_SECRET='your_api_secret_here'")
        print("="*60)
    
    def _log_auth_event(self, event, user, details=""):
        """Log authentication events for security audit"""
        timestamp = datetime.now().isoformat()
        log_entry = f"{timestamp} - {event} - User: {user} - {details}\n"
        
        try:
            with open('api_security.log', 'a') as f:
                f.write(log_entry)
        except:
            pass  # Don't fail if logging fails
    
    def _make_secure_request(self, method, endpoint, **kwargs):
        """
        Make secure API request with proper error handling and logging
        """
        if self.auth_method != "session" and self.auth_method != "token":
            raise Exception("Not authenticated. Use login_with_credentials() or authenticate_with_token()")
        
        # Add security headers
        if 'headers' not in kwargs:
            kwargs['headers'] = {}
        
        # Add timestamp for audit
        kwargs['headers']['X-Request-Time'] = datetime.now().isoformat()
        
        # Make request
        try:
            response = self.session.request(method, f"{self.base_url}{endpoint}", **kwargs)
            
            # Log request for audit
            self._log_request(method, endpoint, response.status_code)
            
            # Handle authentication errors
            if response.status_code == 401:
                print("❌ Authentication failed. Token may be expired.")
                return None
            elif response.status_code == 403:
                print("❌ Access forbidden. Check permissions.")
                return None
            elif response.status_code == 429:
                print("❌ Rate limit exceeded. Please wait.")
                return None
            
            response.raise_for_status()
            return response.json()
            
        except requests.exceptions.RequestException as e:
            print(f"❌ Request failed: {e}")
            self._log_request(method, endpoint, 0, str(e))
            raise
    
    def _log_request(self, method, endpoint, status_code, error=""):
        """Log API requests for audit"""
        timestamp = datetime.now().isoformat()
        user = getattr(self, 'current_user', 'unknown')
        log_entry = f"{timestamp} - {method} {endpoint} - {status_code} - User: {user} - {error}\n"
        
        try:
            with open('api_requests.log', 'a') as f:
                f.write(log_entry)
        except:
            pass
    
    # Secure API methods
    def get(self, endpoint, params=None):
        """Secure GET request"""
        return self._make_secure_request('GET', endpoint, params=params)
    
    def post(self, endpoint, data=None):
        """Secure POST request"""
        return self._make_secure_request('POST', endpoint, json=data)
    
    def put(self, endpoint, data=None):
        """Secure PUT request"""
        return self._make_secure_request('PUT', endpoint, json=data)
    
    def delete(self, endpoint):
        """Secure DELETE request"""
        return self._make_secure_request('DELETE', endpoint)
    
    def logout(self):
        """Logout and clear session"""
        if self.auth_method == "session":
            try:
                self.session.post(f"{self.base_url}/api/method/logout")
                print("✅ Logged out successfully")
            except:
                pass
        
        self.session.cookies.clear()
        self.auth_method = None
        print("🔒 Session cleared")

def demo_secure_usage():
    """
    Demonstrate secure API usage patterns
    """
    print("ERPNext Secure API Client Demo")
    print("="*40)
    
    client = ERPNextSecureClient()
    
    # Method 1: API Token (Recommended for APIs)
    print("\n🔐 Method 1: API Token Authentication (Recommended)")
    print("-" * 50)
    
    if client.authenticate_with_token():
        # Demo secure API calls
        try:
            print("\n📊 Fetching system info...")
            system_settings = client.get('/api/resource/System%20Settings/System%20Settings')
            if system_settings:
                data = system_settings.get('data', {})
                print(f"   Country: {data.get('country', 'Not set')}")
                print(f"   Time Zone: {data.get('time_zone', 'Not set')}")
            
            print("\n👥 Fetching users (limited)...")
            users = client.get('/api/resource/User', params={'limit_page_length': 3})
            if users:
                for user in users.get('data', []):
                    print(f"   - {user.get('full_name', 'Unknown')} ({user.get('name', 'unknown')})")
            
            print("\n🏢 Checking companies...")
            companies = client.get('/api/resource/Company')
            if companies:
                for company in companies.get('data', []):
                    print(f"   - {company.get('name', 'Unknown Company')}")
            
        except Exception as e:
            print(f"❌ Error during API calls: {e}")
        
        client.logout()
    else:
        client.generate_api_key_instructions()
        
        # Method 2: Session Authentication (for web apps)
        print("\n🌐 Method 2: Session Authentication (Web Apps)")
        print("-" * 50)
        print("Would you like to try session-based login? (y/n): ", end='')
        if input().lower().startswith('y'):
            if client.login_with_credentials():
                try:
                    # Demo with session
                    users = client.get('/api/resource/User', params={'limit_page_length': 1})
                    if users:
                        print("✅ Session-based API call successful")
                    
                except Exception as e:
                    print(f"❌ Session API call failed: {e}")
                
                client.logout()

def security_recommendations():
    """
    Print security recommendations
    """
    print("\n" + "="*60)
    print("🔒 SECURITY RECOMMENDATIONS")
    print("="*60)
    print("1. ✅ USE API TOKENS for server-to-server communication")
    print("2. ✅ USE HTTPS in production (never HTTP)")
    print("3. ✅ STORE credentials in environment variables")
    print("4. ✅ IMPLEMENT rate limiting")
    print("5. ✅ LOG all API access for audit trails")
    print("6. ✅ ROTATE API keys regularly (every 90 days)")
    print("7. ✅ USE IP whitelisting when possible")
    print("8. ✅ IMPLEMENT proper error handling")
    print("9. ✅ VALIDATE all inputs")
    print("10. ✅ MONITOR for unusual access patterns")
    print("\n❌ AVOID:")
    print("- Never commit API keys to version control")
    print("- Never use Basic Auth in production")
    print("- Never use HTTP in production")
    print("- Never expose API keys in logs")
    print("- Never use session cookies for mobile apps")
    print("="*60)

if __name__ == "__main__":
    try:
        demo_secure_usage()
        security_recommendations()
    except KeyboardInterrupt:
        print("\n\n👋 Goodbye!")
    except Exception as e:
        print(f"\n❌ Unexpected error: {e}")